### PR TITLE
build empty answers in grading new

### LINKED ISSUE #351: ActionView::Template::Error: undefined method `comment_topic' for nil:NilClass
View details in Rollbar: [https://rollbar.com/jsyeo/FirstProject/items/20/](https://rollbar.com/jsyeo/FirstProject/items/20/)

```

NoMethodError: undefined method `comment_topic' for nil:NilClass
  File "/home/raymoond/jenkins/workspace/coursemology_second/app/views/assessment/gradings/_single_question_form.html.erb", line 90, in block in _app_views_assessment_gradings__single_question_form_html_erb___2124185605930089329_82024140
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/helpers/capture_helper.rb", line 40, in block in capture
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/helpers/capture_helper.rb", line 187, in with_output_buffer
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/haml-4.0.5/lib/haml/helpers/action_view_xss_mods.rb", line 5, in with_output_buffer_with_haml_xss
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/helpers/capture_helper.rb", line 40, in capture
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/haml-4.0.5/lib/haml/helpers/action_view_mods.rb", line 52, in capture_with_haml
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/helpers/form_helper.rb", line 378, in form_for
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/haml-4.0.5/lib/haml/helpers/action_view_mods.rb", line 139, in form_for_with_haml
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/haml-4.0.5/lib/haml/helpers/action_view_xss_mods.rb", line 28, in form_for_with_haml_xss
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/bundler/gems/simple_form-a43bd2722f95/lib/simple_form/action_view_extensions/form_helper.rb", line 29, in block in simple_form_for
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/bundler/gems/simple_form-a43bd2722f95/lib/simple_form/action_view_extensions/form_helper.rb", line 48, in with_simple_form_field_error_proc
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/bundler/gems/simple_form-a43bd2722f95/lib/simple_form/action_view_extensions/form_helper.rb", line 28, in simple_form_for
  File "/home/raymoond/jenkins/workspace/coursemology_second/app/views/assessment/gradings/_single_question_form.html.erb", line 3, in _app_views_assessment_gradings__single_question_form_html_erb___2124185605930089329_82024140
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/template.rb", line 145, in block in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications.rb", line 125, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/template.rb", line 143, in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/partial_renderer.rb", line 265, in render_partial
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/partial_renderer.rb", line 238, in block in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/abstract_renderer.rb", line 38, in block in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications.rb", line 123, in block in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications/instrumenter.rb", line 20, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications.rb", line 123, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/abstract_renderer.rb", line 38, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/newrelic_rpm-3.9.1.236/lib/new_relic/agent/instrumentation/rails3/action_controller.rb", line 177, in block in instrument_with_newrelic
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/newrelic_rpm-3.9.1.236/lib/new_relic/agent/method_tracer.rb", line 219, in trace_execution_scoped
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/newrelic_rpm-3.9.1.236/lib/new_relic/agent/instrumentation/rails3/action_controller.rb", line 176, in instrument_with_newrelic
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/partial_renderer.rb", line 237, in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/renderer.rb", line 41, in render_partial
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/renderer.rb", line 15, in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/helpers/rendering_helper.rb", line 24, in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/haml-4.0.5/lib/haml/helpers/action_view_mods.rb", line 12, in render_with_haml
  File "/home/raymoond/jenkins/workspace/coursemology_second/app/views/assessment/gradings/_form.html.erb", line 2, in _app_views_assessment_gradings__form_html_erb__4412192734096177869_82222620
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/template.rb", line 145, in block in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications.rb", line 125, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/template.rb", line 143, in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/partial_renderer.rb", line 265, in render_partial
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/partial_renderer.rb", line 238, in block in render
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/abstract_renderer.rb", line 38, in block in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications.rb", line 123, in block in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications/instrumenter.rb", line 20, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/activesupport-3.2.14/lib/active_support/notifications.rb", line 123, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/abstract_renderer.rb", line 38, in instrument
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/newrelic_rpm-3.9.1.236/lib/new_relic/agent/instrumentation/rails3/action_controller.rb", line 177, in block in instrument_with_newrelic
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/newrelic_rpm-3.9.1.236/lib/new_relic/agent/method_tracer.rb", line 219, in trace_execution_scoped
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/newrelic_rpm-3.9.1.236/lib/new_relic/agent/instrumentation/rails3/action_controller.rb", line 176, in instrument_with_newrelic
  File "/home/raymoond/.rvm/gems/ruby-2.1.2/gems/actionpack-3.2.14/lib/action_view/renderer/partial_renderer.rb", line 237, in render

### PR DESCRIPTION
**Pull request by wangqiang1208**
Fix #351 #368 #370

If students submitted their missions, after that any new questions added to the mission will cause these exceptions. Because in the submitted missions, there is no answers for the the new question.

Fix this by build empty answers for those questions who don't have answers during grading.  It's a workaround, but as far as think it's the simplest way to solve these issues. 

And this is urgent I think.

### CODE REVIEW

File: app/models/assessment/submission.rb [hunk 1]
```diff
@@ -190,6 +190,26 @@ def eval_answer
...
+  def build_empty_answers
+    return if assessment.questions.count == answers.count
+
+    questions_without_answers = assessment.questions.select do |question|
+      question.answers.where(std_course_id: std_course.id, submission_id: self.id).empty?
+    end
+
+    questions_without_answers.each do |question|
+      default_params = { std_course_id: std_course.id, submission_id: id, question_id: question.id, content: '' }
+      case question.as_question_type
```
Comment (jsyeo) [inline app/models/assessment/submission.rb]: Hmmm... Can we do this http://batsov.com/articles/2012/10/14/ruby-tip-number-3-matching-on-an-objects-class-in-a-case-expression/ ?

**Comment by jsyeo** [comment]
Also, can you write a feature test for this issue?

File: app/models/assessment/submission.rb (hunk 1 above)
Comment (wangqiang1208) [inline app/models/assessment/submission.rb]: Sorry, I don't understand what do you mean here, as_question_type is a string, not class

**Comment by wangqiang1208** [comment]
That one is very time consuming, I don't think I have time now >_<

For the feature test you need to: 
1. create a mission
2. create a user student and enrol him in course and do and submit the mission.
3. add a new question in mission
4. do grading.. 

it's not a simple feature leh..

**Comment by wangqiang1208** [comment]
In my local I've spent like a hour to test on course 135 and some other tests on coding question, confirmed that it works then I pushed here.

**Comment by jsyeo** [comment]
> 1. create a mission
> 2. create a user student and enrol him in course and do and submit the mission.

Can mock this, no?

**Comment by jsyeo** [comment]
I am just afraid of regressions. Your manual tests might work for some cases. I just want to be sure.

**Comment by wangqiang1208** [comment]
Yes you are right. but feature test also cannot test for edge cases, And it's really time consuming, it not easy to do mock also.

**Comment by jsyeo** [comment]
Ok. You're right. This is just a hotfix. We can merge this.

**Comment by wangqiang1208** [comment]
Thanks !
